src/crawlee/request_sources/_request_source.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING
+
+from crawlee import Request
+from crawlee._utils.docs import docs_group
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    from crawlee.base_storage_client._models import ProcessedRequest
+
+
+@docs_group('Abstract classes')
+class RequestSource(ABC):
+    """Abstract base class defining the interface for classes that provide access to a read-only stream of requests.
+
+    Request sources are used to manage and provide access to a storage of crawling requests.
+
+    Key responsibilities:
+        - Fetching the next request to be processed.
+        - Reclaiming requests that failed during processing, allowing retries.
+        - Marking requests as successfully handled after processing.
+        - Managing state information such as the total and handled request counts.
+    """
+
+    @abstractmethod
+    async def get_total_count(self) -> int:
+        """Returns an offline approximation of the total number of requests in the source (i.e. pending + handled)."""
+
+    @abstractmethod
+    async def is_empty(self) -> bool:
+        """Returns True if there are no more requests in the source (there might still be unfinished requests)."""
+
+    @abstractmethod
+    async def is_finished(self) -> bool:
+        """Returns True if all requests have been handled."""
+
+    @abstractmethod
+    async def fetch_next_request(self) -> Request | None:
+        """Returns the next request to be processed, or `null` if there are no more pending requests."""
+
+    @abstractmethod
+    async def mark_request_as_handled(self, request: Request) -> ProcessedRequest | None:
+        """Marks a request as handled after a successful processing (or after giving up retrying)."""
+
+    @abstractmethod
+    async def reclaim_request(self, request: Request, *, forefront: bool = False) -> ProcessedRequest | None:
+        """Reclaims a failed request back to the source, so that it can be returned for processing later again.
+
+        It is possible to modify the request data by supplying an updated request as a parameter.
+        """
```

Comment:
Is there a good reason to keep this one here or should we move it to the `RequestProvider`? Since it requires support for updates to the `Request`, I'd prefer to see it in the provider. It would also make implementing a custom `RequestSource` easier.

Reply:
Hmm, that makes sense to some extent. On the other hand, I feel that `mark_request_as_handled` and `reclaim_request` should be at the same level. You either do one or the other when processing a request, correct? Or maybe both should be moved?

Reply:
Well, if you don't need to support retries, you can do just fine with only `mark_request_as_handled`, and such use case makes some sense.

Reply:
Fetching the next request and marking it as completed later makes it possible to implement a minimal request loader that can survive a migration, which will be a requirement for both the fully fledged `RequestList` and the `SitemapLoader` or whatever we decide to call it.